src/planner/binder/expression/bind_columnref_expression.cpp

```diff
@@ -34,21 +34,49 @@ unique_ptr<ParsedExpression> ExpressionBinder::QualifyColumnName(const string &c
 			return move(coalesce);
 		}
 	}
-	// no table name: find a binding that contains this
+
+	// these could be std::optional given cpp17?
+	unique_ptr<ParsedExpression> matched_from_macro(nullptr);
+	unique_ptr<ParsedExpression> matched_from_alias(nullptr);
+	unique_ptr<ParsedExpression> matched_from_column(nullptr);
+
+	// This preserves existing behaviour.
+	// TODO: change behaviour to error on ambiguity.
 	if (binder.macro_binding != nullptr && binder.macro_binding->HasMatchingBinding(column_name)) {
-		// priority to macro parameter bindings TODO: throw a warning when this name conflicts
 		D_ASSERT(!binder.macro_binding->alias.empty());
-		return make_unique<ColumnRefExpression>(column_name, binder.macro_binding->alias);
-	} else {
+		matched_from_macro = make_unique<ColumnRefExpression>(column_name, binder.macro_binding->alias);
+	}
+
+	if (binder.column_alias_binder && binder.column_alias_binder->HasAlias(column_name)) {
+		unique_ptr<ParsedExpression> real_expression;
+		auto alias_error = binder.column_alias_binder->GetRealExpression(column_name, real_expression);
+		if (!alias_error.empty()) {
+			throw new BinderException(
+			    StringUtil::Format("Alias binding should have worked, but failed: %s", alias_error));
+		}
+		matched_from_alias = move(real_expression);
+	}
+
+	{
 		string table_name = binder.bind_context.GetMatchingBinding(column_name);
-		if (table_name.empty()) {
-			auto similar_bindings = binder.bind_context.GetSimilarBindings(column_name);
-			string candidate_str = StringUtil::CandidatesMessage(similar_bindings, "Candidate bindings");
-			error_message =
-			    StringUtil::Format("Referenced column \"%s\" not found in FROM clause!%s", column_name, candidate_str);
-			return nullptr;
+		if (!table_name.empty()) {
+			matched_from_column = binder.bind_context.CreateColumnReference(table_name, column_name);
 		}
-		return binder.bind_context.CreateColumnReference(table_name, column_name);
+	}
+
+	if (matched_from_macro) {
```

Comment:
I don't think this should be added to the `ExpressionBinder`, but rather this should be specifically added to the `SelectBinder`. Referencing aliases in the other clauses vs referencing them in the `SELECT` clause is different because of *when* the computation of values occurs. In the `HAVING`/`QUALIFY`/`GROUP_BY` clauses we can simply refer to a previously generated value, whereas in the `SELECT` clause we need to duplicate the logic.

Adding this only to the `SelectBinder` will also simplify the code/passing of parameters and arguments.

Reply:
o rly.. this is topical to the mess I'm making locally now. I thought this would let me remove the special casing of alias lookups in `having`/`qualify`/`group by`/`where`/`order by` - the only diff between those is priority of alias vs column for order by (and maybe qualify).
This would have been a nice little cleanup.
Are you saying this isn't workable / has performance implications?
So far I have seen only `group by` "consciously" references the select col. If I'm not wrong, `where`/`having`/`qualify` currently duplicate logic by resolving aliases before binding, I thought there was computation de-duping later on that made that feasible.

Reply:
`WHERE` does need to duplicate the logic since it logically happens before the select, but `HAVING` and `QUALIFY` should not duplicate the logic. There is no cross-operator common subexpression elimination optimizer, so this could have performance implications if this contains complex expressions.

More importantly, this also has potential effects on correctness in case expressions with side-effects are used (e.g. `RANDOM()`). Looking at this, it seems the current implementation is actually incorrect for `WHERE` because of this:

```sql
D select random() a where a>0.5;
┌─────────────────────┐
│          a          │
├─────────────────────┤
│ 0.12791333929219392 │
└─────────────────────┘
```

In the case of `WHERE` and `SELECT` the expression duplication should verify that the expression has no side effects prior to copying (and throw an error if it does). For `HAVING` and `QUALIFY`, since this happens after the projection, it should be possible to just refer to the expression in the projection which solves that problem.

Reply:
I think HAVING and QUALIFY also show the same problem currenly.

Reply:
Ideally `HAVING` and `QUALIFY` could share the same logic, and refer back to the original expression (similar to the grouping). `WHERE` and `SELECT` could also share the same logic, where they would copy the expression (and throw an exception if the expression has side-effects).

Reply:
This is pushing me back to an earlier idea of creating a concrete `binding` in the bind context to refer to the current select, and resolving aliases that way

Reply:
example of group by cleanup: https://github.com/duckdb/duckdb/commit/435c998fd1d4d25b18f7a1b2dd891239e4a4af26

I think this one might be OK regardless of HAVING/QUALIFY, because TryBindGroup seems to achieves de-duping for us. But it would be good to be able to remove special alias handling in the other places as well.

Reply:
more stripping of alias code - i think this is everything but order by.
https://github.com/akdor1154/duckdb/compare/refer_to_select_alias...akdor1154:refer_to_select_alias_working

Even if the current Just-Resolve-Aliases-Before-Binding isn't workable, I still reckon a similar global solution should be able to just handle aliases once instead of having dedicated code in each operator. I'll look back at this in another couple of days.